plugins/generator-1.18.1/forge-1.18.1/templates/tool.java.ftl

```diff
@@ -0,0 +1,426 @@
+		} else {
+			world.playSound(null, entity.getX(), entity.getY(), entity.getZ(), SoundEvents.FISHING_BOBBER_THROW, SoundSource.NEUTRAL, 0.5f, 0.4f / (world.getRandom().nextFloat() * 0.4f + 0.8f));
+			if (!world.isClientSide()) {
+				int k = EnchantmentHelper.getFishingSpeedBonus(itemstack);
+				int j = EnchantmentHelper.getFishingLuckBonus(itemstack);
+				world.addFreshEntity(new FishingHook(entity, world, j, k) {
+
+					@Override public boolean shouldStopFishing(Player player) {
+						if (!player.isRemoved() && player.isAlive() &&
+								(player.getMainHandItem().is(${JavaModName}Items.${data.getModElement().getRegistryNameUpper()}) ||
+								player.getOffhandItem().is(${JavaModName}Items.${data.getModElement().getRegistryNameUpper()}))
+								&& !(this.distanceToSqr(player) > 1024)) {
+							return false;
+						} else {
+							this.discard();
+							return true;
+						}
+					}
+
+				});
+			}
+
+			entity.awardStat(Stats.ITEM_USED.get(this));
+			world.gameEvent(entity, GameEvent.FISHING_ROD_CAST, entity);
+		}
+		
+		<#if hasProcedure(data.onRightClickedInAir)>
+		double x = entity.getX();
+		double y = entity.getY();
+		double z = entity.getZ();
+		<@procedureOBJToCode data.onRightClickedInAir/>
+		</#if>
+
+		return InteractionResultHolder.sidedSuccess(itemstack, world.isClientSide());
+	}
+
+    <@commonMethods/>
+}
+</#if>
+
+<#macro commonMethods>
+	<#if data.stayInGridWhenCrafting>
+        @Override public boolean hasContainerItem(ItemStack stack) {
+        	return true;
+        }
+
+        <#if data.damageOnCrafting && data.usageCount != 0>
+        	@Override public ItemStack getContainerItem(ItemStack itemstack) {
+        		ItemStack retval = new ItemStack(this);
+        		retval.setDamageValue(itemstack.getDamageValue() + 1);
+        		if(retval.getDamageValue() >= retval.getMaxDamage()) {
+        			return ItemStack.EMPTY;
+        		}
+        		return retval;
+        	}
+
+        	@Override public boolean isRepairable(ItemStack itemstack) {
+        		return false;
+        	}
+        <#else>
+        	@Override public ItemStack getContainerItem(ItemStack itemstack) {
+        		return new ItemStack(this);
+        	}
+
+        	<#if data.usageCount != 0>
+        	    @Override public boolean isRepairable(ItemStack itemstack) {
+        		    return false;
+        	    }
+        	</#if>
+        </#if>
+    </#if>
+
+    <#if data.specialInfo?has_content>
+    	@Override public void appendHoverText(ItemStack itemstack, Level world, List<Component> list, TooltipFlag flag) {
+    		super.appendHoverText(itemstack, world, list, flag);
+    		<#list data.specialInfo as entry>
+    		list.add(new TextComponent("${JavaConventions.escapeStringForJava(entry)}"));
+    		</#list>
+    	}
+    </#if>
+
+    <@onItemUsedOnBlock data.onRightClickedOnBlock/>
+
+	<@onCrafted data.onCrafted/>
+
+	<@onEntitySwing data.onEntitySwing/>
+
+	<@onStoppedUsing data.onStoppedUsing/>

```

Comment:
This doesn't work (at least) on shears/fishing rods. Did you test it on items (since it's not tool-specific trigger)? Did it work for you?

Reply:
I did not test yet, just looked through the code.

It doesn't work in 1.16.5 too?

Reply:
I get class version error when downgrading to 1.16.5. But I tested 1.17.1 and it fails too. I could open a bug report for this since it happens for multiple MC versions (I already have all the info on my hands, including demo workspace).

Reply:
I suggest you instead try to open PR fixing it :)

Reply:
Why instead?
Issue + PR )

Reply:
Contributors are generally more welcome to fix the problem when found, not just report it.

Reply:
From what I saw in the code, the entity holding given item prevents this method from being triggered if the world is client side. I wanted to check if there's a difference on server side, but I don't have permissions to change my game mode manually and I can't use (unsupported yet) procedure block to change it programmatically. I think this has no bugs :/

Reply:
Can you test it with some other trigger, eg. place block?

Reply:
That's not an item trigger

Reply:
I know, I meant to place a block on these triggers to verify the work